pkg/apis/sessions/session_state.go

```diff
@@ -256,3 +274,148 @@ func DecodeSessionState(v string, c *encryption.Cipher) (*SessionState, error) {
 	}
 	return ss, nil
 }
+
+// CompressedSessionState returns an lz4 compression of a MessagePack encoded session
+// Encryption & Base64 encoding are delegated to downstream consumers of this method.
+func (s *SessionState) CompressedSessionState() ([]byte, error) {
+	var (
+		ss  SessionState
+		err error
+
+		// Base64 Decode JWTs (if possible)
+		decodedAccessToken  []byte
+		decodedIDToken      []byte
+		decodedRefreshToken []byte
+
+		// LZ4 & MessagePack
+		packed []byte
+		reader *bytes.Reader
+		buf    *bytes.Buffer
+		zw     *lz4.Writer
+	)
+
+	ss = *s
+
+	// We assume if a token successfully base64 decodes, we can encode back later
+	// Otherwise we leave the original token
+	if ss.AccessToken != "" {
+		decodedAccessToken, err = base64.URLEncoding.DecodeString(ss.AccessToken)
+		if err == nil {
+			ss.AccessToken = ""
+		} else {
+			decodedAccessToken = []byte{}
+		}
+	}
+	if ss.IDToken != "" {
+		decodedIDToken, err = base64.URLEncoding.DecodeString(ss.IDToken)
+		if err == nil {
+			ss.IDToken = ""
+		} else {
+			decodedIDToken = []byte{}
+		}
+	}
+	if ss.RefreshToken != "" {
+		decodedRefreshToken, err = base64.URLEncoding.DecodeString(ss.RefreshToken)
+		if err == nil {
+			ss.RefreshToken = ""
+		} else {
+			decodedRefreshToken = []byte{}
+		}
+	}
+
+	// Embed SessionState, Decoded Tokens and Expires pointers into SessionStateCompressed
+	ssc := &SessionStateCompressed{
+		SessionState:        &ss,
+		DecodedAccessToken:  decodedAccessToken,
+		DecodedIDToken:      decodedIDToken,
+		DecodedRefreshToken: decodedRefreshToken,
+	}
+	if !ss.CreatedAt.IsZero() {
+		ssc.CreatedAt = &ss.CreatedAt
+	}
+	if !ss.ExpiresOn.IsZero() {
+		ssc.ExpiresOn = &ss.ExpiresOn
+	}
+
+	//Marshal & Compress the SessionStateCompressed
+	packed, err = msgpack.Marshal(ssc)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// The Compress:Decompress ratio is 1:Many. LZ4 gives fastest decompress speeds
```

Comment:
LZ4 Compression here is the layer that potentially adds the most CPU & speed concerns. LZ4 chosen for its dominance in decompression throughput compared to other decryption libraries (https://github.com/lz4/lz4#benchmarks)

Without LZ4 (but with other tricks), test sessions reduced from ~5000 bytes to ~3500 bytes

With LZ4, reduced further to about ~2700 bytes. No significant difference in our payloads between level 0 compression vs level 9, so left as 0 (fastest)